pkg/apiserver/debugapi/service.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debugapi
+
+import (
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/joomcode/errorx"
+
+	"github.com/pingcap/tidb-dashboard/pkg/apiserver/debugapi/schema"
+	"github.com/pingcap/tidb-dashboard/pkg/apiserver/user"
+)
+
+var (
+	ErrNS      = errorx.NewNamespace("error.api.debugapi")
+	ErrFromDst = ErrNS.NewType("err_from_dst")
+)
+
+func registerRouter(r *gin.RouterGroup, auth *user.AuthService, s *Service) {
+	endpoint := r.Group("/debugapi")
+	endpoint.Use(auth.MWAuthRequired())
+
+	endpoint.GET("/proxy", s.Proxy)
+	endpoint.GET("/endpoint", s.GetEndpointList)
+}
+
+type Service struct {
+	proxy *proxy
+}
+
+func newService() *Service {
+	p := newProxy()
+	for _, endpoint := range schema.EndpointAPIList {
+		p.SetupEndpoint(endpoint)
+	}
+
+	return &Service{proxy: p}
+}
+
+// @Summary Proxy request to tidb/tikv/tiflash/pd http api
+// @Security JwtAuth
+// @Success 200 {object} string
+// @Failure 400 {object} utils.APIError "Bad request"
+// @Failure 401 {object} utils.APIError "Unauthorized failure"
+// @Failure 500 {object} utils.APIError
+// @Router /debugapi/proxy [get]
+func (s *Service) Proxy(c *gin.Context) {
```

Comment:
I would prefer it to be something like `/debugapi/request [post]` and accept a `RequestOptions` that specify request details, like the destination address and the structural parameters. In short, instead of creating a proxy that accepts anything and do some "dangerous" proxy verifying, create an endpoint that accepts structural and safe data and then assemble it and send requests. Constructing from a structural data is always better than filtering a constructed thing.